Metalama.Framework.Engine/SyntaxGeneration/SyntaxGeneratorForIType.TypeSyntaxGeneratorVisitor.cs

```diff
@@ -0,0 +1,299 @@
+                {
+                    // If the inner array we just moved to is also nullable, then
+                    // we must terminate the digging now so we produce the syntax for that,
+                    // and then append the ranks we passed through at the end. This is because
+                    // nullability annotations acts as a "barrier" where we won't reorder array
+                    // through. So whereas:
+                    //
+                    //     string[][,]
+                    //
+                    // is really an array of rank 1 that has an element of rank 2,
+                    //
+                    //     string[]?[,]
+                    //
+                    // is really an array of rank 2 that has nullable elements of rank 1.
+
+                    break;
+                }
+            }
+
+            var elementTypeSyntax = this._declarationSyntaxGenerator.TypeExpression( underlyingType );
+            var ranks = new List<ArrayRankSpecifierSyntax>();
+
+            var arrayType = type;
+            while ( arrayType != null && !arrayType.Equals( underlyingType ) )
+            {
+                ranks.Add( SyntaxFactory.ArrayRankSpecifier(
+                    SyntaxFactory.SeparatedList( Enumerable.Repeat<ExpressionSyntax>( SyntaxFactory.OmittedArraySizeExpression(), arrayType.Rank ) ) ) );
+
+                arrayType = arrayType.ElementType as IArrayType;
+            }
+
+            TypeSyntax arrayTypeSyntax = SyntaxFactory.ArrayType( elementTypeSyntax, SyntaxFactory.List( ranks ) );
+
+            if ( type.IsNullableReferenceType() )
+            {
+                arrayTypeSyntax = SyntaxFactory.NullableType( arrayTypeSyntax );
+            }
+
+            return this.AddInformationTo( arrayTypeSyntax, type );
+        }
+
+        public override TypeSyntax VisitDynamicType( IDynamicType type )
+            => this.AddInformationTo( SyntaxFactory.IdentifierName( "dynamic" ), type );
+
+        public override TypeSyntax VisitFunctionPointerType( IFunctionPointerType functionPointerType ) => throw new NotImplementedException();
+
+        public TypeSyntax CreateSimpleTypeSyntax( INamedType type )
+        {
+            var syntax = this.TryCreateSpecializedNamedTypeSyntax( type );
+            if ( syntax != null )
+            {
+                return syntax;
+            }
+
+            if ( type.Name == string.Empty )
+            {
+                return CreateSystemObject();
+            }
+
+            if ( type.TypeParameters.Count == 0 )
+            {
+                if ( type.TypeKind == TypeKind.Error && type.Name == "var" )
+                {
+                    return CreateSystemObject();
+                }
+
+                return ToIdentifierName( type.Name );
+            }
+
+            var typeArguments = type.IsCanonicalGenericInstance
+                ? Enumerable.Repeat( (TypeSyntax) SyntaxFactory.OmittedTypeArgument(), type.TypeArguments.Count )
```

Comment:
What happens if you want to use the type you introduced as a parameter of it's own method?
`class X<T> { void Foo(X<T> x) {} }`